File: guest-orders-not-linked-to-accounts/python/link_guest_orders.py
"""Link guest WooCommerce orders to the account that shares the same email.

A guest checkout never sets order.customer_id, even when the billing email
matches a real, registered customer. The order sits at customer_id 0 forever,
so it never shows up in "My account", loyalty points never accrue, and any
per-customer report undercounts that shopper. This walks recent guest orders,
looks up a customer by billing email through the WooCommerce REST API, and
confirms the order was really paid by checking the saved Stripe PaymentIntent
before relinking it. Safe to run again and again. Dry run by default.
"""
import os
import logging
import stripe
import requests
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("link_guest_orders")

stripe.api_key = os.environ["STRIPE_SECRET_KEY"]
WOO_URL = os.environ["WOO_STORE_URL"].rstrip("/")
AUTH = HTTPBasicAuth(os.environ["WOO_CONSUMER_KEY"], os.environ["WOO_CONSUMER_SECRET"])
LOOKBACK_DAYS = int(os.environ.get("LOOKBACK_DAYS", "30"))
REQUIRE_PAID = os.environ.get("REQUIRE_PAID", "true").lower() == "true"
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

PAID_STATUSES = {"processing", "completed"}


def intent_id_of(order):
    """The saved Stripe PaymentIntent id, from meta _stripe_intent_id or transaction_id."""
    for meta in order.get("meta_data") or []:
        if meta.get("key") == "_stripe_intent_id" and meta.get("value"):
            return meta["value"]
    tid = order.get("transaction_id")
    return tid if tid and tid.startswith("pi_") else None


def order_amount_minor(order):
    return round(float(order["total"]) * 100)


def decide(order, customers, intent=None):
    """Pure decision. customers is the list returned for the billing email lookup."""
    if order.get("customer_id", 0):
        return ("skip", "order is already linked to an account")
    email = (order.get("billing") or {}).get("email")
    if not email:
        return ("skip", "no billing email on the order")
    if REQUIRE_PAID and order["status"] not in PAID_STATUSES:
        return ("skip", "order is not paid yet")
    if not customers:
        return ("no_account", "no registered account uses this email")
    if len(customers) > 1:
        return ("ambiguous", "more than one account uses this email")
    if REQUIRE_PAID:
        if intent is None:
            return ("unverified", "no Stripe PaymentIntent saved on the order")
        if intent.get("status") != "succeeded":
            return ("unverified", "Stripe does not show this payment as succeeded")
        if abs(order_amount_minor(order) - intent.get("amount_received", 0)) > 1:
            return ("unverified", "order total does not match the Stripe charge")
    return ("link", f"billing email matches account {customers[0]['id']}")


def get_intent(intent_id):
    if not intent_id:
        return None
    try:
        return stripe.PaymentIntent.retrieve(intent_id)
    except stripe.error.InvalidRequestError:
        return None


def find_customers_by_email(email):
    r = requests.get(
        f"{WOO_URL}/wp-json/wc/v3/customers",
        params={"email": email, "per_page": 10},
        auth=AUTH, timeout=30,
    )
    r.raise_for_status()
    return r.json()


def guest_orders():
    page = 1
    after = f"{__import__('datetime').date.today() - __import__('datetime').timedelta(days=LOOKBACK_DAYS)}T00:00:00"
    while True:
        r = requests.get(
            f"{WOO_URL}/wp-json/wc/v3/orders",
            params={"customer": 0, "after": after, "per_page": 50, "page": page},
            auth=AUTH, timeout=30,
        )
        r.raise_for_status()
        batch = r.json()
        if not batch:
            return
        for order in batch:
            yield order
        page += 1


def link_order(order_id, customer_id):
    requests.put(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}",
        json={"customer_id": customer_id},
        auth=AUTH, timeout=30,
    ).raise_for_status()
    requests.post(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}/notes",
        json={"note": f"Linked this guest order to account {customer_id} because the "
                      f"billing email matched a registered customer. Linked by the reconciler."},
        auth=AUTH, timeout=30,
    ).raise_for_status()


def run():
    linked = 0
    for order in guest_orders():
        email = (order.get("billing") or {}).get("email")
        customers = find_customers_by_email(email) if email else []
        intent = get_intent(intent_id_of(order)) if REQUIRE_PAID else None
        action, reason = decide(order, customers, intent)
        if action == "link":
            customer_id = customers[0]["id"]
            log.info("Order %s: %s. %s", order["id"], reason, "would link" if DRY_RUN else "linking")
            if not DRY_RUN:
                link_order(order["id"], customer_id)
            linked += 1
        elif action in ("ambiguous", "unverified"):
            log.warning("Order %s not linked: %s", order["id"], reason)
    log.info("Done. %d order(s) %s.", linked, "to link" if DRY_RUN else "linked")


if __name__ == "__main__":
    run()

File: guest-orders-not-linked-to-accounts/node/link-guest-orders.js
/**
 * Link guest WooCommerce orders to the account that shares the same email.
 *
 * A guest checkout never sets order.customer_id, even when the billing email
 * matches a real, registered customer. The order sits at customer_id 0 forever,
 * so it never shows up in "My account", loyalty points never accrue, and any
 * per-customer report undercounts that shopper. This walks recent guest orders,
 * looks up a customer by billing email through the WooCommerce REST API, and
 * confirms the order was really paid by checking the saved Stripe PaymentIntent
 * before relinking it. Safe to run again and again. Dry run by default.
 *
 * Guide: https://www.allanninal.dev/woocommerce/guest-orders-not-linked-to-accounts/
 */
import Stripe from "stripe";
import { pathToFileURL } from "node:url";

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || "sk_test_dummy");
const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const LOOKBACK_DAYS = Number(process.env.LOOKBACK_DAYS || 30);
const REQUIRE_PAID = (process.env.REQUIRE_PAID || "true").toLowerCase() === "true";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const PAID_STATUSES = new Set(["processing", "completed"]);

export function intentIdOf(order) {
  for (const meta of order.meta_data || []) {
    if (meta.key === "_stripe_intent_id" && meta.value) return meta.value;
  }
  const tid = order.transaction_id;
  return tid && tid.startsWith("pi_") ? tid : null;
}

export function orderAmountMinor(order) {
  return Math.round(parseFloat(order.total) * 100);
}

/** Pure decision. customers is the list returned for the billing email lookup. */
export function decide(order, customers, intent = null) {
  if (order.customer_id) return ["skip", "order is already linked to an account"];
  const email = (order.billing || {}).email;
  if (!email) return ["skip", "no billing email on the order"];
  if (REQUIRE_PAID && !PAID_STATUSES.has(order.status)) return ["skip", "order is not paid yet"];
  if (!customers || customers.length === 0) return ["no_account", "no registered account uses this email"];
  if (customers.length > 1) return ["ambiguous", "more than one account uses this email"];
  if (REQUIRE_PAID) {
    if (!intent) return ["unverified", "no Stripe PaymentIntent saved on the order"];
    if (intent.status !== "succeeded") return ["unverified", "Stripe does not show this payment as succeeded"];
    if (Math.abs(orderAmountMinor(order) - (intent.amount_received || 0)) > 1) {
      return ["unverified", "order total does not match the Stripe charge"];
    }
  }
  return ["link", `billing email matches account ${customers[0].id}`];
}

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  if (!res.ok) throw new Error(`Woo ${path} returned ${res.status}`);
  return res.json();
}

async function getIntent(intentId) {
  if (!intentId) return null;
  try {
    return await stripe.paymentIntents.retrieve(intentId);
  } catch {
    return null;
  }
}

async function findCustomersByEmail(email) {
  return woo(`/customers?email=${encodeURIComponent(email)}&per_page=10`);
}

async function* guestOrders() {
  const after = new Date(Date.now() - LOOKBACK_DAYS * 86400000).toISOString();
  let page = 1;
  while (true) {
    const batch = await woo(`/orders?customer=0&after=${after}&per_page=50&page=${page}`);
    if (!batch.length) return;
    for (const order of batch) yield order;
    page++;
  }
}

async function linkOrder(orderId, customerId) {
  await woo(`/orders/${orderId}`, {
    method: "PUT",
    body: JSON.stringify({ customer_id: customerId }),
  });
  await woo(`/orders/${orderId}/notes`, {
    method: "POST",
    body: JSON.stringify({
      note: `Linked this guest order to account ${customerId} because the billing email ` +
            `matched a registered customer. Linked by the reconciler.`,
    }),
  });
}

export async function run() {
  let linked = 0;
  for await (const order of guestOrders()) {
    const email = (order.billing || {}).email;
    const customers = email ? await findCustomersByEmail(email) : [];
    const intent = REQUIRE_PAID ? await getIntent(intentIdOf(order)) : null;
    const [action, reason] = decide(order, customers, intent);
    if (action === "link") {
      const customerId = customers[0].id;
      console.log(`Order ${order.id}: ${reason}. ${DRY_RUN ? "would link" : "linking"}`);
      if (!DRY_RUN) await linkOrder(order.id, customerId);
      linked++;
    } else if (action === "ambiguous" || action === "unverified") {
      console.warn(`Order ${order.id} not linked: ${reason}`);
    }
  }
  console.log(`Done. ${linked} order(s) ${DRY_RUN ? "to link" : "linked"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
